Add erc20 docs (#689)
OpenZeppelin/cairo-contracts: docs/modules/ROOT/pages/erc20.adoc

    ) -> bool;
    fn approve(spender: ContractAddress, amount: u256) -> bool;
}
----

=== ERC20 compatibility

Although Starknet is not EVM compatible, this implementation aims to be as close as possible to the ERC20 standard, in the following ways:

* It uses Cairo's `u256` instead of `felt252`.
* It makes use of Cairo's short strings to simulate `name` and `symbol`.

But some differences can still be found, such as:

* Function selectors are written in snake case as opposed to camel case in Solidity.
* `transfer`, `transfer_from` and `approve` will never return anything different from `true` because they will revert on any error.

== Extensibility

:components: https://community.starknet.io/t/cairo-1-contract-syntax-is-evolving/94794#extensibility-and-components-11[Components]

WARNING: This is currently unsafe, because storage members could clash among used contracts if not reviewed carefully. Extensibility will be revisited after {components} are introduced.

ERC20 contracts can be extended by utilizing the contract's `_unsafe_new_contract_state()`.
The unsafe contract state allows access to the `ERC20Impl`.

> This enables `IERC20` methods to include additional behaviors.

Reviewer: I think is better to have the warning below the paragraph, because I think is not clear what "This" is referring to when we haven't presented the unsafe mechanism.

I would also change the wording from "This is currently unsafe, because storage members..." to "This mechanism is unsafe because storage members..."

[,javascript]
----
#[external(v0)]
fn transfer(self: @ContractState, recipient: ContractAddress, amount: u256) -> bool {
    let unsafe_state = ERC20::unsafe_new_contract_state();

    // Add custom logic here

    ERC20::ERC20Impl::transfer(@unsafe_state, recipient, amount)
}
----

The unsafe state also allows access to ERC20's `InternalImpl`.
Having access to ERC20's internal implementation permits contracts, for example, to expose the internal methods like `_mint`.

[,javascript]
----
#[external(v0)]
fn mint(self: @ContractState, to: ContractAddress, amount: u256) {
    let unsafe_state = ERC20::unsafe_new_contract_state();

>     assert_only_owner();  // Ensure the mint function is protected!

Reviewer: ```suggestion
    
    // Ensures the mint function is protected!
    assert_only_owner(); 
```

== Extensibility

:components: https://community.starknet.io/t/cairo-1-contract-syntax-is-evolving/94794#extensibility-and-components-11[Components]

ERC20 contracts can be extended by utilizing the contract's `_unsafe_new_contract_state()`.
The unsafe contract state allows access to the `ERC20Impl`.
This enables `IERC20` methods to include additional behaviors.

[,javascript]
----
#[external(v0)]
fn transfer(self: @ContractState, recipient: ContractAddress, amount: u256) -> bool {
    let unsafe_state = ERC20::unsafe_new_contract_state();

    // Add custom logic here

    ERC20::ERC20Impl::transfer(@unsafe_state, recipient, amount)
}
----

The unsafe state also allows access to ERC20's `InternalImpl`.
Having access to ERC20's internal implementation permits contracts, for example, to expose the internal methods like `_mint`.

[,javascript]
----
#[external(v0)]
fn mint(self: @ContractState, to: ContractAddress, amount: u256) {
    let unsafe_state = ERC20::unsafe_new_contract_state();
    
    // Ensures the mint function is protected!
    assert_only_owner(); 
    ERC20::InternalImpl::_mint(@unsafe_state, to, amount);
}
----

> WARNING: This mechanism is unsafe because storage members could clash among used contracts if not reviewed carefully. Extensibility will be revisited after {components} are introduced.

Author: I changed it because there's more context to the warning after the content. Seeing it both ways, I agree it's better to have it in the beginning since it's harder to miss

== Presets

> The ERC20 contract itself includes a basic implementation preset and can be used as-is for quick prototyping and testing.

Author: That's fair. The idea was to have the intro paragraph with the expectation of more presets to come (I think it would less redundant then). I'll remove

Extensibility will be revisited after {components} are introduced.

Using Contracts for Cairo, constructing an ERC20 contract requires setting up the constructor and exposing the ERC20 interface.
Here's what that looks like:

[,javascript]
----
#[starknet::contract]
mod MyToken {
    use starknet::ContractAddress;
    use openzeppelin::token::erc20::ERC20;

    #[storage]
    struct Storage {}

    #[constructor]
    fn constructor(
        self: @ContractState,
        initial_supply: u256,
        recipient: ContractAddress
    ) {
        let name = 'MyToken';
        let symbol = 'MTK';

        let mut unsafe_state = ERC20::unsafe_new_contract_state();
        ERC20::InternalImpl::initializer(ref unsafe_state, name, symbol);
        ERC20::InternalImpl::_mint(ref unsafe_state, recipient, initial_supply);
    }
}

    #[external(v0)]
    impl MyTokenImpl of IERC20<ContractState> {
        fn name(self: @ContractState) -> felt252 {
            let mut unsafe_state = ERC20::unsafe_new_contract_state();
            ERC20::ERC20Impl::name(ref unsafe_state, name, symbol); 
        }

>         ...

Reviewer: ```suggestion
        (...)
```

I think this is easier to read, and I've been using this in the code examples. If you think is better without parenthesis, let me know to update in my side.

Reviewer: cc @martriay

Author: I lean toward no parenthesis but no strong opinions if we want to include them

Reviewer 2: +1 to no parenthesis, +1 no strong opinion

Author: I ended up adding in the parenthesis because they do look a little naked with all empty space

Reviewer: Just to confirm, should we use them then, right?

Author: Yes, sir. At least for now

    ) -> bool;
    fn approve(spender: ContractAddress, amount: u256) -> bool;
}
----

=== ERC20 compatibility

:cairo-selectors: https://github.com/starkware-libs/cairo/blob/7dd34f6c57b7baf5cd5a30c15e00af39cb26f7e1/crates/cairo-lang-starknet/src/contract.rs#L39-L48[Cairo]
:solidity-selectors: https://solidity-by-example.org/function-selector/[Solidity]

Although Starknet is not EVM compatible, this implementation aims to be as close as possible to the ERC20 standard, in the following ways:

* It makes use of Cairo's short strings to simulate `name` and `symbol`.

But some differences can still be found, such as:

* Function selectors are written in snake_case as opposed to camelCase in Solidity.
* `transfer`, `transfer_from` and `approve` will never return anything different from `true` because they will revert on any error.
* Function selectors are calculated differently between {cairo-selectors} and {solidity-selectors}.

== Usage

:components: https://community.starknet.io/t/cairo-1-contract-syntax-is-evolving/94794#extensibility-and-components-11[Components]
:erc20-supply: xref:/guides/erc20-supply.adoc[Creating ERC20 Supply]

> WARNING: The following example uses a contract's `unsafe_new_contract_state` to access another contract's state.
> This is currently unsafe, because storage members could clash among used contracts if not reviewed carefully.
> Extensibility will be revisited after {components} are introduced.

Reviewer 2: ```suggestion
WARNING: The following example uses `unsafe_new_contract_state` to access another contract's state.
Although this is useful to use them as modules, it's considered unsafe because storage members could clash among used contracts if not reviewed carefully.
Extensibility will be revisited after {components} are introduced.
```
i think this reads better. that being said, is it really unsafe? if we pass `src5_storage()` to `SRC5` functions, is it problematic? I think the issue arises if there's storage collisions among different used modules, but that's mitigated by prefixes. either way, it's not related to `unsafe_new_contract_state`, isn't the same with components?

Author: Great questions! The warning was added a while ago. As you mentioned, prefixes mitigate the potential for storage clashes. AFAIK that's what's unsafe about it. I'll remove the warning

Until we have components, though, I don't think it's a bad idea to inform users of this...it probably belongs in the forthcoming `Extensibility` doc

Reviewer 2: sure we can inform about this. what do you think about the suggestion?

Author: I think the suggestion indeed reads better!

Just to clarify, do we agree to remove the warning now and insert it into `Extensibility` in the future? If not, should we include the warning in each doc `Usage` section for consistency?

Reviewer 2: ideally we would add a single explanation in Extensibility and then link there from every section, but i wonder whether it makes sense to write such a page for unsafe storage usage when components are around the corner

Author: Hmm right. I don't like the idea of assuming components will arrive soon. How's this: let's make a tentative plan to include an unsafe storage section in Extensibility. If we get components sooner, we'll have to make adjustments to Extensibility and the other docs anyway. If we don't have components by the time the other docs are done, we can just carry on until tenemos componentes

* <<extensibility,Extensibility>>
* <<presets,Presets>>
 ** <<erc20_basic,ERC20 (basic)>>
* <<api_specification,API Specification>>
 ** <<erc20impl,ERC20Impl>>
  *** <<name,`name`>>
  *** <<symbol,`symbol`>>
  *** <<decimals,`decimals`>>
  *** <<total_supply,`total_supply`>>
  *** <<balance_of,`balance_of`>>
  *** <<allowance,`allowance`>>
  *** <<transfer,`transfer`>>
  *** <<transfer_from,`transfer_from`>>
  *** <<approve,`approve`>>
 ** <<events,Events>>
  *** <<Transfer,`Transfer`>>
  *** <<Approval,`Approval`>>
 ** <<non-standard-external,Non-standard external>>
  *** <<increase_allowance,`increase_allowance`>>
  *** <<decrease_allowance,`decrease_allowance`>>
 ** <<internalimpl,InternalImpl>>
  *** <<initializer,initializer>>
  *** <<internal_increase_allowance,_increase_allowance>>
  *** <<internal_decrease_allowance,_decrease_allowance>>
  *** <<internal_mint,_mint>>
  *** <<internal_burn,_burn>>
  *** <<internal_approve,_approve>>
  *** <<internal_transfer,_transfer>>
  *** <<internal_spend_allowance,_spend_allowance>>

> == Interface

Reviewer: I think the rest of the methods (like `increase_allowance`) should appear here somehow as well. I don't think we should document the ERC20 interface, but our implementation interface (maybe specifying what methods are the ERC20 standard and what are not, but all of them should appear documented IMO).

Author: > I think the rest of the methods (like `increase_allowance`) should appear here somehow as well.

I put the non standard methods in the `Non-standard external` section (just under `Events`)—unless I'm misunderstanding your comment

> I don't think we should document the ERC20 interface, but our implementation interface (maybe specifying what methods are the ERC20 standard and what are not, but all of them should appear documented IMO)

I had a similar thought with just documenting our implementation and insert the `ERC20ABI` as the introductory interface. I'm open to the change if you really want to push for it. The OZ-contracts docs includes the [IERC20](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20) and the [ERC20](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#ERC20) implementation interface. I suppose we could include something like `IERC20 standard method that...` in the appropriate method descriptions

Reviewer 2: as in the solidity-docs, `IERC20` and `ERC20ABI` are properly documented in the API section. i agree with Eric's comment on the main section interface. We should document our module (today it's a preset used a module, tomorrow it'll be a component), and mention it implements `IERC20` and exposes `ERC20ABI`.

[,rust]
----

> #[starknet::interface]
> trait IERC20<TState> {
>     fn name(self: @TState) -> felt252;
>     fn symbol(self: @TState) -> felt252;
>     fn decimals(self: @TState) -> u8;
>     fn total_supply(self: @TState) -> u256;
>     fn balance_of(self: @TState, account: ContractAddress) -> u256;
>     fn allowance(self: @TState, owner: ContractAddress, spender: ContractAddress) -> u256;
>     fn transfer(ref self: TState, recipient: ContractAddress, amount: u256) -> bool;
>     fn transfer_from(
>         ref self: TState, sender: ContractAddress, recipient: ContractAddress, amount: u256
>     ) -> bool;
>     fn approve(ref self: TState, spender: ContractAddress, amount: u256) -> bool;

Reviewer: I think this may be easier to read (in the UI)

```suggestion
#[starknet::interface]
trait IERC20<TState> {
    fn name(self: @TState) -> felt252;
    fn symbol(self: @TState) -> felt252;
    fn decimals(self: @TState) -> u8;
    fn total_supply(self: @TState) -> u256;
    fn balance_of(self: @TState, account: ContractAddress) -> u256;
    fn allowance(
        self: @TState,
        owner: ContractAddress,
        spender: ContractAddress
    ) -> u256;
    fn transfer(
        ref self: TState, 
        recipient: ContractAddress, 
        amount: u256
    ) -> bool;
    fn transfer_from(
        ref self: TState, 
        sender: ContractAddress, 
        recipient: ContractAddress, 
        amount: u256
    ) -> bool;
    fn approve(ref self: TState, spender: ContractAddress, amount: u256) -> bool;
```

}
----

=== ERC20 compatibility

Although Starknet is not EVM compatible, this implementation aims to be as close as possible to the ERC20 standard, in the following ways:

* It uses Cairo's `u256` instead of `felt252`.
* It makes use of Cairo's short strings to simulate `name` and `symbol`.

But some differences can still be found, such as:

* `transfer`, `transfer_from` and `approve` will never return anything different from `true` because they will revert on any error.

> * Function selectors are calculated differently between https://github.com/starkware-libs/cairo/blob/7dd34f6c57b7baf5cd5a30c15e00af39cb26f7e1/crates/cairo-lang-starknet/src/contract.rs#L39-L48[Cairo] and https://solidity-by-example.org/function-selector/[Solidity].

Reviewer: I'm not sure we should include this here. The fact that function selectors are computed differently is not directly related to the implementation of the ERC20 standard, because it says nothing about selectors, right? And also, using a different casing would give different selectors anyway. Casing may be included too.

Author: That's fair. The selector calculation is more of a network difference, not really specific to ERC20. And yeah, good call on including the casing difference

== Usage

Use cases go from medium of exchange currency to voting rights, staking, and more.

Coming soon!

== Extensibility

> (*This is currently unsafe*)

Reviewer: ```suggestion
:components: https://community.starknet.io/t/cairo-1-contract-syntax-is-evolving/94794#extensibility-and-components-11[Components]

WARNING: This is currently unsafe, because storage members could clash among used contracts if not reviewed carefully. Extensibility will be revisited after {components} are introduced.
```

Reviewer: <img width="765" alt="Screenshot 2023-08-14 at 13 27 35" src="https://github.com/OpenZeppelin/cairo-contracts/assets/26070224/2f2a3b8d-34ae-4251-80b6-fa412b89cbc3">

* Function selectors are written in snake case as opposed to camel case in Solidity.
* `transfer`, `transfer_from` and `approve` will never return anything different from `true` because they will revert on any error.

== Usage

Use cases go from medium of exchange currency to voting rights, staking, and more.

Coming soon!

== Extensibility

:components: https://community.starknet.io/t/cairo-1-contract-syntax-is-evolving/94794#extensibility-and-components-11[Components]

WARNING: This is currently unsafe, because storage members could clash among used contracts if not reviewed carefully. Extensibility will be revisited after {components} are introduced.

ERC20 contracts can be extended by utilizing the contract's _unsafe_new_contract_state().
The unsafe contract state allows access to the _ERC20Impl_.
This enables IERC20 methods to include additional behaviors.

[,rust]
----
#[external(v0)]
fn transfer(self: @ContractState, recipient: ContractAddress, amount: u256) -> bool {
    let unsafe_state = ERC20::unsafe_new_contract_state();

    // Add custom logic here

    ERC20::ERC20Impl::transfer(@unsafe_state, recipient, amount)
}
----

> The unsafe state also allows access to ERC20's _InternalImpl_.

Reviewer: ```suggestion
The unsafe state also allows access to ERC20's `InternalImpl`.
```

> Having access to ERC20's interal implementation permits contracts, for example, to expose the internal methods like _mint.

Reviewer: ```suggestion
Having access to ERC20's internal implementation permits contracts, for example, to expose the internal methods like `_mint`.
```

[,rust]
----
#[external(v0)]
fn mint(self: @ContractState, to: ContractAddress, amount: u256) {
    let unsafe_state = ERC20::unsafe_new_contract_state();
    assert_only_owner();  // Ensure the mint function is protected!
    ERC20::InternalImpl::_mint(@unsafe_state, to, amount);
}
----

> Note that accessing the contract's state in this way is unsafe at the moment.

Reviewer: I think we can remove this line having the warning on top of the section.

== Presets

The ERC20 contract itself includes a basic implementation preset and can be used as-is for quick prototyping and testing.
The basic preset mints an inital supply which is necessary because it does not expose a minting function.

More presets coming soon!

=== ERC20 (basic)

*TODO: CHANGE LINK TO RELEASE COMMIT WHEN AVAILABLE*

The https://github.com/OpenZeppelin/cairo-contracts/blob/cairo-2/src/token/erc20/erc20.cairo[`ERC20`] preset offers a quick and easy setup for deploying a basic ERC20 token.

== API Specification

=== ERC20Impl

> [,rust]

Reviewer: ```suggestion
[,javascript]
```

----
fn name(self: @TState) -> felt252;
fn symbol(self: @TState) -> felt252;
fn decimals(self: @TState) -> u8;
fn total_supply(self: @TState) -> u256;
fn balance_of(self: @TState, account: ContractAddress) -> u256;
fn allowance(
    self: @TState,
    owner: ContractAddress,
    spender: ContractAddress
) -> u256;
fn transfer(
    ref self: TState,
    recipient: ContractAddress,
    amount: u256
) -> bool;
fn transfer_from(
    ref self: TState,
    sender: ContractAddress,
    recipient: ContractAddress,
    amount: u256
) -> bool;

> fn approve(

Reviewer: Notes about things I can't select:

- Change StarkNet for Starknet (line 4).

    ref self: TState,
    spender: ContractAddress,
    amount: u256
) -> bool;
----

> ==== `name`

Reviewer: We can remove this headers. I added them for the names to appear in the right-side menu, but in this case we don't have that. I think we can follow the [solidity docs](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20-totalSupply--) version for this, without the return keyword as well.

Reviewer: This applies for the rest of the methods as well.

> [.contract-item]

Reviewer: ```suggestion
[.contract-item#name]
```

This allows the TOC to continue linking the section.

==== `[.contract-item-name]#++name++#++(self: @TState) → felt252++`

Returns the name of the token.

> ===== Returns

Reviewer: We can remove this section here, and in the rest of the methods to match [solidity docs](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20-totalSupply--), because we are not documenting a CLI or NRE as in Nile, but a contract.

== Interface

:dual-interfaces: xref:/interfaces.adoc#dual_interfaces[Dual Interfaces]
:erc20-supply: xref:/guides/erc20-supply.adoc[Creating ERC20 Supply]
:version: https://github.com/OpenZeppelin/cairo-contracts/releases/tag/v0.7.0[Contracts v0.7.0]
:custom-decimals: xref:/erc20.adoc#customizing_decimals[Customizing Decimals]

The following interface represents the full interface of the Contracts for Cairo `ERC20ABI` module.
The interface includes the `IERC20` standard interface as well as the non-standard functions e.g. `increase_allowance`.
To support older deployments of ERC20, as mentioned in {dual-interfaces}, `ERC20ABI` also includes the previously defined functions written in camelCase.

> WARNING: Prior to {version}, ERC20 iterations store and read `decimals` from storage; however, this library's implementation returns a static `18`.

Reviewer 2: why is this under `interface`? shouldn't it be either under the decimals section, or above at the beginning?

But some differences can still be found, such as:

> * The standard function selectors are written in snake_case as opposed to camelCase in Solidity.

Reviewer 2: why the standard ones? fwiw we have both casings, should we mention that too?
